File: javascript/constrainedGA.js
class ConstrainedGA {
    constructor({ objFunc, inequalityConstraints, equalityConstraints, bounds, 
                  K = 200, iterMax = 200, p_s = 0.1, p_m = 0.05, p_l = 0.01, lam = 1000.0, maxTries = 10 }) {
        this.f = objFunc;
        this.g = inequalityConstraints || [];
        this.h = equalityConstraints || [];
        this.bounds = bounds; // Format: [[min1, max1], [min2, max2], ...]
        this.n = bounds.length;

        // Hyperparameters from the paper (Table 1)
        this.K = K;
        this.iterMax = iterMax;
        this.p_s = p_s;
        this.p_m = p_m;
        this.p_l = p_l;
        this.lam = lam; // Base penalty multiplier parameter \lambda
        this.maxTries = maxTries;
        
        // Function evaluation tracker
        this.fevals = 0; 
    }

    // Checks if a chromosome satisfies boundary envelopes and constraints
    isFeasible(x) {
        for (let j = 0; j < this.n; j++) {
            if (x[j] < this.bounds[j][0] || x[j] > this.bounds[j][1]) return false;
        }
        for (let gi of this.g) {
            if (gi(x) > 1e-5) return false;
        }
        for (let hj of this.h) {
            if (Math.abs(hj(x)) > 1e-5) return false;
        }
        return true;
    }

    // Section 2.1: Fitness Evaluation using Adaptive Penalty Technique
    evaluateFitness(x, currentGeneration = 100) {
        this.fevals++; // Increment evaluation counter
        
        const v1 = this.f(x);
        
        // v2: Equality constraint violations (squared sum) [Eq. 3]
        const v2 = this.h.reduce((sum, hj) => sum + Math.pow(hj(x), 2), 0);
        
        // v3: Inequality constraint violations where G(x) = max(0, x) [Eq. 4 & 5]
        const v3 = this.g.reduce((sum, gi) => sum + Math.pow(Math.max(0, gi(x)), 2), 0);
        
        // Dynamic adaptive multiplier scale: Escalates penalty weight over generational progress
        // to forcefully reject infeasible individuals out of highly restrictive search spaces.
        let adaptiveLam = this.lam;
        if (v2 > 1e-5 || v3 > 1e-5) {
            adaptiveLam = this.lam * (1 + currentGeneration * 0.5);
        }
        
        // Final penalized fitness function mapping [Eq. 6]
        return v1 + adaptiveLam * v2 + adaptiveLam * v3;
    }

    // Step 3: Initialize K chromosomes inside the bounds
    initializePopulation() {
        const population = [];
        let attempts = 0;
        const maxAttempts = this.K * 10; 

        // Attempt 1: Try to pull strictly feasible spaces if problem configuration allows easily
        while (population.length < this.K && attempts < maxAttempts) {
            attempts++;
            let x = [];
            for (let j = 0; j < this.n; j++) {
                const [min, max] = this.bounds[j];
                x.push(min + Math.random() * (max - min));
            }
            if (this.isFeasible(x)) {
                population.push(x);
            }
        }

        // Attempt 2: Instant Fallback to complete initialization immediately
        while (population.length < this.K) {
            let x = [];
            for (let j = 0; j < this.n; j++) {
                const [min, max] = this.bounds[j];
                x.push(min + Math.random() * (max - min));
            }
            population.push(x);
        }
        
        return population;
    }

    // Tournament Selection helper for crossover
    tournamentSelection(population, fitnesses) {
        const idx1 = Math.floor(Math.random() * this.K);
        const idx2 = Math.floor(Math.random() * this.K);
        return fitnesses[idx1] < fitnesses[idx2] ? population[idx1] : population[idx2];
    }

    // Algorithm 2: Crossover Procedure
    crossover(population, fitnesses) {
        // 1. Sort chromosomes based on fitness (best to worst)
        const indices = Array.from({ length: this.K }, (_, i) => i);
        indices.sort((a, b) => fitnesses[a] - fitnesses[b]);
        let sortedPop = indices.map(i => population[i]);

        // 3. Create children
        const children = [];
        const numLoops = Math.floor(((1 - this.p_s) * this.K) / 2);

        for (let i = 0; i < numLoops; i++) {
            const p1 = this.tournamentSelection(sortedPop, fitnesses);
            const p2 = this.tournamentSelection(sortedPop, fitnesses);

            let c1 = [], c2 = [];
            for (let j = 0; j < this.n; j++) {
                const a = -0.5 + Math.random() * 2; // Random scale alpha in [-0.5, 1.5]
                c1.push(a * p1[j] + (1 - a) * p2[j]);
                c2.push(a * p2[j] + (1 - a) * p1[j]);
                
                // Keep coordinates within explicit boundaries
                c1[j] = Math.max(this.bounds[j][0], Math.min(this.bounds[j][1], c1[j]));
                c2[j] = Math.max(this.bounds[j][0], Math.min(this.bounds[j][1], c2[j]));
            }
            children.push(c1, c2);
        }

        // 5. Replace worst individuals if feasibility states align
        const replaceLimit = Math.floor((1 - this.p_s) * this.K);
        for (let i = 0; i < replaceLimit; i++) {
            if (i >= children.length) break;
            const child = children[i];
            const replaceIdx = this.K - i - 1;

            const pFeas = this.isFeasible(sortedPop[replaceIdx]);
            const cFeas = this.isFeasible(child);

            if ((pFeas && cFeas) || (!pFeas && !cFeas)) {
                sortedPop[replaceIdx] = child;
            }
        }
        return sortedPop;
    }

    // Algorithm 4: Procedure perturb(x)
    perturb(x) {
        let newX = [...x];
        for (let j = 0; j < this.n; j++) {
            if (Math.random() <= this.p_m) {
                const xi = Math.random();
                newX[j] = this.bounds[j][0] + xi * (this.bounds[j][1] - this.bounds[j][0]);
            }
        }
        return newX;
    }

    // Algorithm 5: Procedure perturbFeasible(x)
    perturbFeasible(x) {
        let newX = [...x];
        for (let j = 0; j < this.n; j++) {
            if (Math.random() <= this.p_m) {
                let tries = 0;
                const xOld = newX[j];
                
                while (tries < this.maxTries) {
                    const xi = Math.random();
                    newX[j] = this.bounds[j][0] + xi * (this.bounds[j][1] - this.bounds[j][0]);
                    
                    if (this.isFeasible(newX)) {
                        break;
                    } else {
                        newX[j] = xOld;
                    }
                    tries++;
                }
            }
        }
        return newX;
    }

    // Algorithm 3: Mutation Procedure
    mutation(population) {
        // Skip index 0 to preserve elite structural chromosome
        for (let i = 1; i < this.K; i++) {
            if (this.isFeasible(population[i])) {
                population[i] = this.perturbFeasible(population[i]);
            } else {
                population[i] = this.perturb(population[i]);
            }
        }
        return population;
    }

    // Section 2.3: Adaptive Coordinate Pattern Search (COBYLA Function Substitute)
    localSearch(population, currentGeneration) {
        for (let i = 0; i < this.K; i++) {
            if (Math.random() <= this.p_l) {
                let current = [...population[i]];
                let currentFit = this.evaluateFitness(current, currentGeneration);
                
                // Adaptive step allocation: narrow search window for high dimensions
                let stepSize = this.n > 4 ? 0.01 : 0.1; 
                const contractionFactor = 0.5;
                const tolerance = 1e-6; 

                while (stepSize > tolerance) {
                    let improved = false;

                    for (let j = 0; j < this.n; j++) {
                        // Test positive coordinate drift
                        let candidatePositive = [...current];
                        candidatePositive[j] = Math.max(this.bounds[j][0], Math.min(this.bounds[j][1], current[j] + stepSize));
                        
                        if (this.isFeasible(candidatePositive)) {
                            let fitPos = this.evaluateFitness(candidatePositive, currentGeneration);
                            if (fitPos < currentFit) {
                                current = candidatePositive;
                                currentFit = fitPos;
                                improved = true;
                                continue;
                            }
                        }

                        // Test negative coordinate drift
                        let candidateNegative = [...current];
                        candidateNegative[j] = Math.max(this.bounds[j][0], Math.min(this.bounds[j][1], current[j] - stepSize));
                        
                        if (this.isFeasible(candidateNegative)) {
                            let fitNeg = this.evaluateFitness(candidateNegative, currentGeneration);
                            if (fitNeg < currentFit) {
                                current = candidateNegative;
                                currentFit = fitNeg;
                                improved = true;
                            }
                        }
                    }

                    // Contract vector steps if local coordinates fail optimization
                    if (!improved) {
                        stepSize *= contractionFactor;
                    }
                }
                population[i] = current;
            }
        }
        return population;
    }

    // Statistical mathematical variance processing helper
    calculateVariance(arr) {
        const mean = arr.reduce((a, b) => a + b, 0) / arr.length;
        return arr.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / arr.length;
    }

    // Main Run Step Orchestration Loop (Algorithm 1)
    solve() {
        this.fevals = 0; // Reset runtime counter metric
        let population = this.initializePopulation();
        let iterCount = 0;

        let bestHistory = []; 
        const windowSize = 15; // Generation window frame observing variance flatlining

        while (iterCount < this.iterMax) {
            // Step 4: Evaluate generation fitness
            const fitnesses = population.map(chrom => this.evaluateFitness(chrom, iterCount));

            // Select global optimum candidate
            let minFitIdx = 0;
            for (let i = 1; i < this.K; i++) {
                if (fitnesses[i] < fitnesses[minFitIdx]) minFitIdx = i;
            }
            const currentBestVal = fitnesses[minFitIdx];
            bestHistory.push(currentBestVal);

            // --- Section 2.4: Asymptotic Variance History Stopping Rule ---
            if (iterCount >= windowSize) {
                const recentBestValues = bestHistory.slice(-windowSize);
                const currentVariance = this.calculateVariance(recentBestValues);
                
                // Explicit breakout condition mapping flat progression
                if (currentVariance <= 1e-11) {
                    break; 
                }
            }

            // Step 5 & 6: Vector Genetic Operations and Stochastic Exploitation Local Step
            population = this.crossover(population, fitnesses);
            population = this.mutation(population);
            population = this.localSearch(population, iterCount);

            iterCount++;
        }

        // Run absolute final evaluation cleanup array maps
        const finalFitnesses = population.map(chrom => this.evaluateFitness(chrom, iterCount));
        
        let bestIdx = 0;
        let bestIsFeasible = this.isFeasible(population[bestIdx]);

        for (let i = 1; i < this.K; i++) {
            const currentIsFeasible = this.isFeasible(population[i]);

            // Feasibility-First Rule:
            // 1. If this candidate is valid but our current best is invalid, the valid one wins instantly.
            // 2. If both share the same feasibility state, choose the one with the better fitness score.
            if (currentIsFeasible && !bestIsFeasible) {
                bestIdx = i;
                bestIsFeasible = true;
            } else if (currentIsFeasible === bestIsFeasible) {
                if (finalFitnesses[i] < finalFitnesses[bestIdx]) {
                    bestIdx = i;
                }
            }
        }

        return {
            coordinates: population[bestIdx],
            value: this.f(population[bestIdx]),
            fevalsSpent: this.fevals 
        };
    }
}

if (typeof module !== 'undefined') {
    module.exports = ConstrainedGA;
}

File: matlab/ConstrainedGA.m
classdef ConstrainedGA < handle
    % ConstrainedGA - Genetic Algorithm with Adaptive Penalty for Constrained Optimization
    %
    % This class implements a constrained genetic algorithm using adaptive
    % penalty techniques, tournament selection, crossover, mutation, and
    % local search for solving optimization problems with constraints.
    %
    % Usage:
    %   solver = ConstrainedGA('ObjectiveFunction', @(x) sum(x.^2), ...
    %                          'Bounds', [0 1; 0 1], ...
    %                          'InequalityConstraints', {@(x) x(1) + x(2) - 1});
    %   result = solver.solve();
    
    properties
        f                   % Objective function handle
        g                   % Cell array of inequality constraint function handles (g(x) <= 0)
        h                   % Cell array of equality constraint function handles (h(x) = 0)
        bounds              % [n x 2] matrix of bounds [min max] for each variable
        n                   % Number of variables
        
        % Hyperparameters
        K                   % Population size (default: 200)
        iterMax             % Maximum iterations (default: 200)
        p_s                 % Survival rate (default: 0.1)
        p_m                 % Mutation probability (default: 0.05)
        p_l                 % Local search probability (default: 0.01)
        lam                 % Base penalty multiplier (default: 1000.0)
        maxTries            % Max perturbation attempts (default: 10)
        
        % Tracking
        fevals              % Function evaluation counter
    end
    
    methods
        function obj = ConstrainedGA(varargin)
            % Constructor - Initialize the Constrained GA
            %
            % Name-Value Pairs:
            %   'ObjectiveFunction' - Function handle for objective @(x)
            %   'InequalityConstraints' - Cell array of function handles {@(x), ...}
            %   'EqualityConstraints' - Cell array of function handles {@(x), ...}
            %   'Bounds' - [n x 2] matrix of [min max] for each variable
            %   'PopulationSize' - Population size K (default: 200)
            %   'MaxGenerations' - Maximum iterations (default: 200)
            %   'SurvivalRate' - Survival rate p_s (default: 0.1)
            %   'MutationRate' - Mutation probability p_m (default: 0.05)
            %   'LocalSearchRate' - Local search probability p_l (default: 0.01)
            %   'PenaltyMultiplier' - Base penalty λ (default: 1000.0)
            %   'MaxTries' - Max perturbation attempts (default: 10)
            
            p = inputParser;
            addParameter(p, 'ObjectiveFunction', []);
            addParameter(p, 'InequalityConstraints', {});
            addParameter(p, 'EqualityConstraints', {});
            addParameter(p, 'Bounds', []);
            addParameter(p, 'PopulationSize', 200);
            addParameter(p, 'MaxGenerations', 200);
            addParameter(p, 'SurvivalRate', 0.1);
            addParameter(p, 'MutationRate', 0.05);
            addParameter(p, 'LocalSearchRate', 0.01);
            addParameter(p, 'PenaltyMultiplier', 1000.0);
            addParameter(p, 'MaxTries', 10);
            parse(p, varargin{:});
            
            obj.f = p.Results.ObjectiveFunction;
            obj.g = p.Results.InequalityConstraints;
            obj.h = p.Results.EqualityConstraints;
            obj.bounds = p.Results.Bounds;
            obj.n = size(obj.bounds, 1);
            
            obj.K = p.Results.PopulationSize;
            obj.iterMax = p.Results.MaxGenerations;
            obj.p_s = p.Results.SurvivalRate;
            obj.p_m = p.Results.MutationRate;
            obj.p_l = p.Results.LocalSearchRate;
            obj.lam = p.Results.PenaltyMultiplier;
            obj.maxTries = p.Results.MaxTries;
            
            obj.fevals = 0;
        end
        
        function feasible = isFeasible(obj, x)
            % Check if chromosome satisfies bounds and constraints
            feasible = true;
            
            % Check bounds
            for j = 1:obj.n
                if x(j) < obj.bounds(j,1) || x(j) > obj.bounds(j,2)
                    feasible = false;
                    return;
                end
            end
            
            % Check inequality constraints g(x) <= 0
            for i = 1:length(obj.g)
                if obj.g{i}(x) > 1e-5
                    feasible = false;
                    return;
                end
            end
            
            % Check equality constraints h(x) = 0
            for i = 1:length(obj.h)
                if abs(obj.h{i}(x)) > 1e-5
                    feasible = false;
                    return;
                end
            end
        end
        
        function fitness = evaluateFitness(obj, x, currentGeneration)
            % Evaluate fitness using adaptive penalty technique
            if nargin < 3
                currentGeneration = 100;
            end
            
            obj.fevals = obj.fevals + 1;
            
            % Objective value
            v1 = obj.f(x);
            
            % Equality constraint violations (squared sum)
            v2 = 0;
            for i = 1:length(obj.h)
                v2 = v2 + obj.h{i}(x)^2;
            end
            
            % Inequality constraint violations
            v3 = 0;
            for i = 1:length(obj.g)
                v3 = v3 + max(0, obj.g{i}(x))^2;
            end
            
            % Adaptive penalty multiplier
            adaptiveLam = obj.lam;
            if v2 > 1e-5 || v3 > 1e-5
                adaptiveLam = obj.lam * (1 + currentGeneration * 0.5);
            end
            
            % Penalized fitness
            fitness = v1 + adaptiveLam * v2 + adaptiveLam * v3;
        end
        
        function population = initializePopulation(obj)
            % Initialize K chromosomes within bounds
            population = cell(obj.K, 1);
            attempts = 0;
            maxAttempts = obj.K * 10;
            popCount = 0;
            
            % Attempt 1: Try to generate feasible solutions
            while popCount < obj.K && attempts < maxAttempts
                attempts = attempts + 1;
                x = obj.bounds(:,1) + rand(obj.n, 1) .* (obj.bounds(:,2) - obj.bounds(:,1));
                if obj.isFeasible(x)
                    popCount = popCount + 1;
                    population{popCount} = x;
                end
            end
            
            % Attempt 2: Fill remaining with random solutions
            while popCount < obj.K
                x = obj.bounds(:,1) + rand(obj.n, 1) .* (obj.bounds(:,2) - obj.bounds(:,1));
                popCount = popCount + 1;
                population{popCount} = x;
            end
        end
        
        function parent = tournamentSelection(obj, population, fitnesses)
            % Tournament selection
            idx1 = randi(obj.K);
            idx2 = randi(obj.K);
            if fitnesses(idx1) < fitnesses(idx2)
                parent = population{idx1};
            else
                parent = population{idx2};
            end
        end
        
        function population = crossover(obj, population, fitnesses)
            % Crossover procedure (Algorithm 2)
            
            % Sort population by fitness
            [~, indices] = sort(fitnesses);
            sortedPop = population(indices);
            
            % Create children
            children = {};
            numLoops = floor(((1 - obj.p_s) * obj.K) / 2);
            
            for i = 1:numLoops
                p1 = obj.tournamentSelection(sortedPop, fitnesses);
                p2 = obj.tournamentSelection(sortedPop, fitnesses);
                
                % Blended crossover
                alpha = -0.5 + rand * 2; % α ∈ [-0.5, 1.5]
                c1 = alpha * p1 + (1 - alpha) * p2;
                c2 = alpha * p2 + (1 - alpha) * p1;
                
                % Enforce bounds
                c1 = max(obj.bounds(:,1), min(obj.bounds(:,2), c1));
                c2 = max(obj.bounds(:,1), min(obj.bounds(:,2), c2));
                
                children{end+1} = c1;
                children{end+1} = c2;
            end
            
            % Replace worst individuals with feasibility awareness
            replaceLimit = floor((1 - obj.p_s) * obj.K);
            for i = 1:min(replaceLimit, length(children))
                child = children{i};
                replaceIdx = obj.K - i + 1;
                
                pFeas = obj.isFeasible(sortedPop{replaceIdx});
                cFeas = obj.isFeasible(child);
                
                if (pFeas && cFeas) || (~pFeas && ~cFeas)
                    sortedPop{replaceIdx} = child;
                end
            end
            
            population = sortedPop;
        end
        
        function newX = perturb(obj, x)
            % Perturb chromosome (Algorithm 4)
            newX = x;
            for j = 1:obj.n
                if rand <= obj.p_m
                    xi = rand;
                    newX(j) = obj.bounds(j,1) + xi * (obj.bounds(j,2) - obj.bounds(j,1));
                end
            end
        end
        
        function newX = perturbFeasible(obj, x)
            % Perturb chromosome while trying to maintain feasibility (Algorithm 5)
            newX = x;
            for j = 1:obj.n
                if rand <= obj.p_m
                    tries = 0;
                    xOld = newX(j);
                    
                    while tries < obj.maxTries
                        xi = rand;
                        newX(j) = obj.bounds(j,1) + xi * (obj.bounds(j,2) - obj.bounds(j,1));
                        
                        if obj.isFeasible(newX)
                            break;
                        else
                            newX(j) = xOld;
                        end
                        tries = tries + 1;
                    end
                end
            end
        end
        
        function population = mutation(obj, population)
            % Mutation procedure (Algorithm 3)
            % Skip index 1 to preserve elite
            for i = 2:obj.K
                if obj.isFeasible(population{i})
                    population{i} = obj.perturbFeasible(population{i});
                else
                    population{i} = obj.perturb(population{i});
                end
            end
        end
        
        function population = localSearch(obj, population, currentGeneration)
            % Adaptive coordinate pattern search
            for i = 1:obj.K
                if rand <= obj.p_l
                    current = population{i};
                    currentFit = obj.evaluateFitness(current, currentGeneration);
                    
                    % Adaptive step size
                    if obj.n > 4
                        stepSize = 0.01;
                    else
                        stepSize = 0.1;
                    end
                    contractionFactor = 0.5;
                    tolerance = 1e-6;
                    
                    while stepSize > tolerance
                        improved = false;
                        
                        for j = 1:obj.n
                            % Test positive direction
                            candidatePos = current;
                            candidatePos(j) = max(obj.bounds(j,1), min(obj.bounds(j,2), current(j) + stepSize));
                            
                            if obj.isFeasible(candidatePos)
                                fitPos = obj.evaluateFitness(candidatePos, currentGeneration);
                                if fitPos < currentFit
                                    current = candidatePos;
                                    currentFit = fitPos;
                                    improved = true;
                                    continue;
                                end
                            end
                            
                            % Test negative direction
                            candidateNeg = current;
                            candidateNeg(j) = max(obj.bounds(j,1), min(obj.bounds(j,2), current(j) - stepSize));
                            
                            if obj.isFeasible(candidateNeg)
                                fitNeg = obj.evaluateFitness(candidateNeg, currentGeneration);
                                if fitNeg < currentFit
                                    current = candidateNeg;
                                    currentFit = fitNeg;
                                    improved = true;
                                end
                            end
                        end
                        
                        % Contract step size if no improvement
                        if ~improved
                            stepSize = stepSize * contractionFactor;
                        end
                    end
                    
                    population{i} = current;
                end
            end
        end
        
        function result = solve(obj)
            % Main optimization loop (Algorithm 1)
            obj.fevals = 0;
            population = obj.initializePopulation();
            iterCount = 0;
            bestHistory = [];
            windowSize = 15;
            
            while iterCount < obj.iterMax
                % Evaluate fitness
                fitnesses = zeros(obj.K, 1);
                for i = 1:obj.K
                    fitnesses(i) = obj.evaluateFitness(population{i}, iterCount);
                end
                
                % Track best value
                [currentBestVal, minFitIdx] = min(fitnesses);
                bestHistory(end+1) = currentBestVal;
                
                % Check convergence
                if iterCount >= windowSize
                    recentBest = bestHistory(end-windowSize+1:end);
                    currentVariance = var(recentBest);
                    if currentVariance <= 1e-11
                        break;
                    end
                end
                
                % Genetic operations
                population = obj.crossover(population, fitnesses);
                population = obj.mutation(population);
                population = obj.localSearch(population, iterCount);
                
                iterCount = iterCount + 1;
            end
            
            % Final evaluation and selection
            finalFitnesses = zeros(obj.K, 1);
            for i = 1:obj.K
                finalFitnesses(i) = obj.evaluateFitness(population{i}, iterCount);
            end
            
            bestIdx = 1;
            bestIsFeasible = obj.isFeasible(population{bestIdx});
            
            for i = 2:obj.K
                currentIsFeasible = obj.isFeasible(population{i});
                
                % Feasibility-first rule
                if currentIsFeasible && ~bestIsFeasible
                    bestIdx = i;
                    bestIsFeasible = true;
                elseif currentIsFeasible == bestIsFeasible
                    if finalFitnesses(i) < finalFitnesses(bestIdx)
                        bestIdx = i;
                    end
                end
            end
            
            % Return results
            result.coordinates = population{bestIdx};
            result.value = obj.f(population{bestIdx});
            result.fevalsSpent = obj.fevals;
            result.isFeasible = obj.isFeasible(population{bestIdx});
        end
    end
end
